// @accounts/server/lib/utils/email.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
exports.emailTemplates = {
    from: 'accounts-js <no-reply@accounts-js.com>',
    verifyEmail: {
        subject: function () { return 'Verify your account email'; },
        text: function (user, url) {
            return "To verify your account email please click on this link: " + url;
        },
        html: function (user, url) {
            return "To verify your account email please <a href=\"" + url + "\">click here</a>.";
        },
    },
    resetPassword: {
        subject: function () { return 'Reset your password'; },
        text: function (user, url) { return "To reset your password please click on this link: " + url; },
        html: function (user, url) {
            return "To reset your password please <a href=\"" + url + "\">click here</a>.";
        },
    },
    enrollAccount: {
        subject: function () { return 'Set your password'; },
        text: function (user, url) { return "To set your password please click on this link: " + url; },
        html: function (user, url) {
            return "To set your password please <a href=\"" + url + "\">click here</a>.";
        },
    },
    passwordChanged: {
        subject: function () { return 'Your password has been changed'; },
        text: function () { return "Your account password has been successfully changed"; },
        html: function () { return "Your account password has been successfully changed."; },
    },
};
exports.sendMail = function (mail) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
        console.warn('No configuration for email, you must set an email configuration');
        console.warn('Documentation: https://accounts-js.netlify.com/docs/email');
        console.log(mail);
        return [2 /*return*/];
    });
}); };
